### PR TITLE
[WIP] add test for tcms.testplans.views.NewTestPlanView.

### LINKED ISSUE #1616: Automated tests for missing coverage in tcms.testplans.views.NewTestPlanView
Kiwi TCMS needs additional automated test(s) as mentioned in subject.

**Definition of Done**:
- full coverage for the module or class mentioned in the title. This means both positive/negative scenarios where this makes sense or scenarios which traverse different branches of the code where that makes sense.
- test class/test method names are descriptive and reflect the behavior of the subject under test.  A good scheme to follow is `Test<ClassUnderTest>.test_[who][what][expected_results]`
- CI pass
- All code review passed & PR merged

Hints:
- Coverage report is at https://codecov.io/gh/kiwitcms/Kiwi/tree/master/tcms
- Devel getting started docs: https://kiwitcms.readthedocs.io/en/latest/contribution.html#testing
- Django testing documentation: https://docs.djangoproject.com/en/3.0/topics/testing/

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by awalvie**
Fix #1616

### CODE REVIEW

File: tcms/testplans/tests/test_views.py [hunk 1]
```diff
@@ -9,6 +9,119 @@
                                   ProductFactory, VersionFactory, PlanTypeFactory)
 
 
+class NewPlanViewTest(LoggedInTestCase):
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:12]: There are some improvements here but generally looks like a copy of `NewPlanViewTest` from `testplans/tests/test_new_plan.py`.  Add your changes (+ the change of the base class) to `test_new_plan.py` so we can review again.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testplans/tests/test_views.py (hunk 1 above)
Comment (awalvie) [reply tcms/testplans/tests/test_views.py:12]: Roger that, I mostly used the changed from the `EditPlanViewTest` didn't know there already was an implementation at `test_new_plan.py`.

File: tcms/testplans/tests/test_new_plan.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 from django.urls import reverse
+from django.utils.translation import gettext_lazy as _
 
 from tcms.testplans.models import TestPlan
 from tcms.testplans.tests.tests import BasePlanTest
 from tcms.tests import remove_perm_from_user, user_should_have_perm
+from tcms.tests.factories import VersionFactory
 
 
 class NewPlanViewTest(BasePlanTest):
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:12]: Can you make this inherit from LoggedInTestCase. Shouldn't need many changes.

File: tcms/testplans/tests/test_new_plan.py [hunk 3]
```diff
@@ -109,3 +111,49 @@ def test_post_with_no_perm_redirects_to_login(self):
         response = self.client.post(self.location, self.request, follow=True)
 
         self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:113]: These scenarios around permissions must inherit from the PermissionsTestCase class instead of being implemented here, see https://github.com/kiwitcms/Kiwi/pull/2076/files

File: tcms/testplans/tests/test_new_plan.py [hunk 4]
```diff
@@ -109,3 +111,49 @@ def test_post_with_no_perm_redirects_to_login(self):
...
+        # Note: version not assigned to the current Product, that's why
+        # it is invalid !
+        version = VersionFactory()
+        new_data['product_version'] = version.pk
+
+        response = self.client.post(self.location, data=new_data, follow=True)
+
+        self.assertContains(
+            response,
+            _('Select a valid choice. That choice is not one of the available choices.'))
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:158]: All of these 3 scenarios are exactly the same - trigger a form error which displays an error. They just do it via different fields.

However this part is handled internally by Django so we can argue that we don't really need to test that b/c Django already does.   At most only 1 of these scenarios should stay if you are keen on keeping them.

File: tcms/testplans/tests/test_new_plan.py [hunk 5]
```diff
@@ -109,3 +111,49 @@ def test_post_with_no_perm_redirects_to_login(self):
...
+
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:159]: However none of the 3 scenarios actually covers the jump from line 50-58 in the view under test b/c that doesn't happen when the form is invalid. It happens when the main form is valid but the inline form for email notifications is invalid. For more hints see https://github.com/kiwitcms/Kiwi/pull/2076/files

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testplans/tests/test_new_plan.py [hunk 6]
```diff
@@ -76,36 +118,117 @@ def test_plan_create_new_inactive(self):
...
+            }
+        )
+
+    def verify_get_with_permission(self):
+        response = self.client.get(self.url)
+
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, _('Create new TestPlan'))
+
+    def verify_post_with_permission(self):
```
Comment (awalvie) [inline tcms/testplans/tests/test_new_plan.py:207]: I can't figure out why this is failing

File: tcms/testplans/tests/test_new_plan.py [hunk 7]
```diff
@@ -1,73 +1,115 @@
...
     def test_plan_new_get(self):
-        response = self.client.get(self.location)
+        response = self.client.get(self.url)
 
         self.assertContains(
             response,
             '<input class="bootstrap-switch" name="is_active" type="checkbox" checked',
-            html=False)
+            html=False,
+        )
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:64]: Why do you need all of these unrelated changes ?

File: tcms/testplans/tests/test_new_plan.py [hunk 8]
```diff
@@ -1,73 +1,115 @@
...
+            "email_settings-TOTAL_FORMS": "1",
+            "email_settings-INITIAL_FORMS": "1",
+            "email_settings-MIN_NUM_FORMS": "0",
+            "email_settings-MAX_NUM_FORMS": "1",
+            "is_active": True,
         }
 
     def test_plan_new_get(self):
-        response = self.client.get(self.location)
+        response = self.client.get(self.url)
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:58]: That's again an unrelated change so why do we need it ?

File: tcms/testplans/tests/test_new_plan.py [hunk 9]
```diff
@@ -1,73 +1,115 @@
...
+
+class TestNewTestPlanView(LoggedInTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        cls.product = ProductFactory()
+        cls.product_version = VersionFactory(product=cls.product)
+        cls.test_plan_type = PlanTypeFactory()
+        cls.test_plan_1 = TestPlanFactory()
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:31]: Bring back `self.plan_name` and remove `cls.test_plan_1`.

File: tcms/testplans/tests/test_new_plan.py [hunk 10]
```diff
@@ -76,36 +118,117 @@ def test_plan_create_new_inactive(self):
         self._test_plan_create_new(is_active=False)
 
     def _test_plan_create_new(self, is_active):
-        self.request['is_active'] = is_active
+        self.request["is_active"] = is_active
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:121]: why do we need this change ?

File: tcms/testplans/tests/test_new_plan.py [hunk 11]
```diff
@@ -76,36 +118,117 @@ def test_plan_create_new_inactive(self):
...
+            name=self.request["name"],
             is_active=is_active,
+            product=self.request["product"],
         )
-        self.assertEqual(plan.author, self.user)
+        self.assertEqual(plan.author, self.tester)
         self.assertEqual(plan.product, self.product)
         self.assertEqual(plan.product_version, self.product_version)
-        self.assertEqual(plan.type, self.plan_type)
+        self.assertEqual(plan.type, self.test_plan_type)
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:134]: Keep the attribute names consistent to minimize unnecessary changes.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testplans/tests/test_new_plan.py (hunk 8 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:58]: I don't quite understand with what you mean by unrelated, is it because I'm not keeping names consistent in both the tests?

File: tcms/testplans/tests/test_new_plan.py (hunk 10 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:121]: I don't quite understand, do you mean why do we need the test or why the double quotes instead of single quotes?

File: tcms/testplans/tests/test_new_plan.py (hunk 11 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:134]: yeah, sorry about that.

File: tcms/testplans/tests/test_new_plan.py (hunk 7 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:64]: I don't quite understand what you mean by unrelated, can you please elaborate? (I'm sorry I'm a little slow to understand things)

Comment (atodorov) [reply tcms/testplans/tests/test_new_plan.py:64]: many of the changes you have made are updating formatting only and are not related to the issue in question nor are adding new automated tests. Their only purpose is to change formatting of existing code.

So as the first step I am asking why do we need them added in this pull request and what necessitates these changes ?

File: tcms/testplans/tests/test_new_plan.py (hunk 8 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:58]: If you're referring to changing the name from location to url, I've done so to keep the name consistent in both tests.

File: tcms/testplans/tests/test_new_plan.py (hunk 7 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:64]: Yeah, about the formatting changes, it's my bad, I accidentally ran black on my changes, which did this.

File: tcms/testplans/tests/test_new_plan.py (hunk 11 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:134]: changing `user` to `tester` is because of inheriting from `PermissionsTestCase`

File: tcms/testplans/tests/test_new_plan.py (hunk 8 above)
Comment (atodorov) [reply tcms/testplans/tests/test_new_plan.py:58]: The two base test classes are different so there isn't much of a consistency between them. Next time keep the changes to minimum b/c it makes following the diff easier.

File: tcms/testplans/tests/test_new_plan.py [hunk 12]
```diff
@@ -1,53 +1,62 @@
...
+
+class TestNewTestPlanView(LoggedInTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        cls.product = ProductFactory()
+        cls.product_version = VersionFactory(product=cls.product)
+        cls.plan_type = PlanTypeFactory()
+        cls.plan_name = TestPlanFactory()
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:31]: `.plan_name` is a string in the original, not an object.

File: tcms/testplans/tests/test_new_plan.py [hunk 13]
```diff
@@ -96,16 +122,96 @@ def _test_plan_create_new(self, is_active):
...
+            name=self.post_data["name"],
+            product=self.post_data["product"],
+        )
+        redirect_url = reverse('test_plan_url', args=[test_plan.pk, slugify(test_plan.name)])
+
+        self.assertRedirects(response, redirect_url)
+        self._assert_test_plan(test_plan)
 
-        response = self.client.get(self.location, follow=True)
+    def verify_get_without_permission(self):
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:201]: this is already handled by PermissionTestCase itself and you rarely have to override it. Comment it out and if the tests pass then it is not needed. The same applies to verify_post_viothout_permission.

File: tcms/testplans/tests/test_new_plan.py [hunk 14]
```diff
@@ -1,53 +1,62 @@
...
         }
 
     def test_plan_new_get(self):
-        response = self.client.get(self.location)
+        response = self.client.get(self.url)
 
-        self.assertContains(
-            response,
-            '<input class="bootstrap-switch" name="is_active" type="checkbox" checked',
-            html=False)
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:50]: Why would you delete this assertion ?

File: tcms/testplans/tests/test_new_plan.py [hunk 15]
```diff
@@ -62,13 +71,29 @@ def test_plan_new_get(self):
             'type="checkbox" checked', html=False)
         self.assertContains(
             response,
-            '<input class="bootstrap-switch" name="email_settings-0-notify_on_plan_update" '
+            '<input class="bootstrap-switch" name="email_settings-0-notify_on_case_update" '
             'type="checkbox" checked', html=False)
         self.assertContains(
             response,
-            '<input class="bootstrap-switch" name="email_settings-0-notify_on_case_update" '
+            '<input class="bootstrap-switch" name="email_settings-0-notify_on_plan_update" '
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:78]: Why swapping the order of the assertions here and on line 74 ?

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testplans/tests/test_new_plan.py (hunk 8 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:58]: alrighty

File: tcms/testplans/tests/test_new_plan.py (hunk 6 above)
Comment (atodorov) [reply tcms/testplans/tests/test_new_plan.py:196]: because self.tester is missing testplans.view_testplan permission

Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:197]: I can't request a review so pinging @atodorov

File: tcms/testplans/tests/test_new_plan.py [hunk 16]
```diff
@@ -96,16 +126,87 @@ def _test_plan_create_new(self, is_active):
...
+        self.assertContains(
+            response,
+            '<input class="bootstrap-switch" name="email_settings-0-auto_to_plan_author" '
+            'type="checkbox"',
+            html=False,
+        )
+
 
-        response = self.client.get(self.location, follow=True)
+class TestNewTestPlanViewPermissions(PermissionsTestCase):
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:151]: pylint: disable=too-many-ancestors

File: tcms/testplans/tests/test_new_plan.py [hunk 17]
```diff
@@ -96,16 +126,87 @@ def _test_plan_create_new(self, is_active):
...
+
+        cls.plan_name = 'plan_name'
+        cls.product = ProductFactory()
+        cls.product_version = VersionFactory(product=cls.product)
+        cls.plan_type = PlanTypeFactory()
+        cls.test_plan = TestPlanFactory()
+
+        user_should_have_perm(cls.tester, perm="testplans.view_testplan")
+
+        cls.post_data.update(
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:170]: pylint: disable=objects-update-used

File: tcms/testplans/tests/test_new_plan.py [hunk 18]
```diff
@@ -1,44 +1,57 @@
...
+        user_should_have_perm(cls.tester, perm="testplans.view_testplan")
+
+        cls.location = reverse("plans-new")
+
+        cls.request = {
+            "author": cls.tester.pk,
+            "product": cls.product.pk,
+            "product_version": cls.product_version.pk,
+            "type": cls.plan_type.pk,
+            "name": cls.plan_name,
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:43]: The rest of the code in this class isn't using the class attribute so this can be hard-coded value. Also can use the class name as the value (or in combination with string format) to give it more uniqueness because in the test class for permissions you are using the same name and then searching by name.

File: tcms/testplans/tests/test_new_plan.py [hunk 19]
```diff
@@ -82,10 +111,11 @@ def _test_plan_create_new(self, is_active):
         self.assertEqual(response.status_code, HTTPStatus.FOUND)
 
         plan = TestPlan.objects.get(
-            name=self.plan_name,
+            name=self.request["name"],
             is_active=is_active,
+            product=self.request["product"],
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:116]: Or alternatively to the previous comment you can keep cls.plan_name, just give it some uniqueness based on the test class name for example and you will not have to filter by product.

File: tcms/testplans/tests/test_new_plan.py [hunk 20]
```diff
@@ -96,16 +126,87 @@ def _test_plan_create_new(self, is_active):
...
+    permission_label = add_testplan_permission
+    http_method_names = ["get", "post"]
+    url = reverse("plans-new")
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.post_data = {"description": "test"}
+        super().setUpTestData()
+
+        cls.plan_name = 'plan_name'
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:162]: Give this value some uniqueness.

File: tcms/testplans/tests/test_new_plan.py [hunk 21]
```diff
@@ -96,16 +126,87 @@ def _test_plan_create_new(self, is_active):
...
+    def verify_post_with_permission(self):
+        response = self.client.post(self.url, self.post_data, follow=True)
+        test_plan = TestPlan.objects.get(
+            name=self.post_data["name"],
+            product=self.post_data["product"],
+        )
+        redirect_url = reverse('test_plan_url', args=[test_plan.pk, slugify(test_plan.name)])
 
-        response = self.client.post(self.location, self.request, follow=True)
+        self.assertRedirects(response, redirect_url)
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:205]: There is also a short URL (/plan/123/) which itself redirects to the one with the slugified name. IDK if assertRedirects will work if you try the short URL but that may save you from using slugify in the tests.

File: tcms/testplans/tests/test_new_plan.py [hunk 22]
```diff
@@ -96,16 +126,87 @@ def _test_plan_create_new(self, is_active):
...
+        self._assert_test_plan(test_plan)
 
-        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+    def _assert_test_plan(self, test_plan):
+        self.assertEqual(test_plan.author, self.tester)
+        self.assertEqual(test_plan.product, self.product)
+        self.assertEqual(test_plan.product_version, self.product_version)
+        self.assertEqual(test_plan.type, self.plan_type)
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:212]: All of these are incorrect b/c you compare the record from the DB with the values which were passed via POST and if the POST request returned a success these will always be the same.

The important behavior is to verify that after the redirect we've actually landed on the page showing the newly created object. assertRedirects above does a good job of it. assertContains on a unique plan name would compliment that and it would be enough. No need to check all fields.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testplans/tests/test_new_plan.py (hunk 21 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py:205]: I did try using that but it doesn't work with the short URL.

File: tcms/testplans/tests/test_new_plan.py (hunk 22 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py]: I've made the changes, pinging because I can't re-request a review @atodorov.

File: tcms/testplans/tests/test_new_plan.py (hunk 16 above)
Comment (atodorov) [reply tcms/testplans/tests/test_new_plan.py]: See falke8

File: tcms/testplans/tests/test_new_plan.py (hunk 17 above)
Comment (atodorov) [reply tcms/testplans/tests/test_new_plan.py]: See falke8

Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py]: fixed!

File: tcms/testplans/tests/test_new_plan.py (hunk 16 above)
Comment (awalvie) [reply tcms/testplans/tests/test_new_plan.py]: fixed!

File: tcms/testplans/tests/test_new_plan.py [hunk 23]
```diff
@@ -96,16 +127,83 @@ def _test_plan_create_new(self, is_active):
...
+    def verify_post_with_permission(self):
+        response = self.client.post(self.url, self.post_data, follow=True)
+        test_plan = TestPlan.objects.get(
+            name=self.post_data["name"],
+        )
+        redirect_url = reverse('test_plan_url', args=[test_plan.pk, slugify(test_plan.name)])
 
-        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+        self.assertRedirects(response, redirect_url)
+        self.assertContains(response, _(self.post_data["name"]))
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:209]: Plan name isn't translated so you don't need to compare it as translated.

File: tcms/testplans/tests/test_new_plan.py [hunk 24]
```diff
@@ -82,10 +113,10 @@ def _test_plan_create_new(self, is_active):
         self.assertEqual(response.status_code, HTTPStatus.FOUND)
 
         plan = TestPlan.objects.get(
-            name=self.plan_name,
+            name=self.request["name"],
```
Comment (atodorov) [inline tcms/testplans/tests/test_new_plan.py:116]: Either use `self.plan_name` or hard-code the value in `self.request`. No need to use both at the same time b/c `plan_name` isn't used anywhere else as-is.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Review by atodorov — APPROVED** [review]
